### PR TITLE
Compatibility injected into channel

### PR DESCRIPTION
**Pull request by pgomulka**
an experiment based on the idea https://github.com/elastic/elasticsearch/pull/60516#issuecomment-685944711

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/node/Node.java [hunk 1]
```diff
@@ -685,6 +688,20 @@ protected Node(final Environment initialEnvironment,
...
+     */
+    private CompatibleVersion getRestCompatibleFunction() {
+        List<RestCompatibilityPlugin> restCompatibilityPlugins = pluginsService.filterPlugins(RestCompatibilityPlugin.class);
+        CompatibleVersion compatibleVersion = CompatibleVersion.CURRENT_VERSION;
+        if (restCompatibilityPlugins.size() > 1) {
+            throw new IllegalStateException("Only one RestCompatibilityPlugin is allowed");
+        } else if (restCompatibilityPlugins.size() == 1) {
+            compatibleVersion = restCompatibilityPlugins.get(0)::getCompatibleVersion;
+        }
+        return compatibleVersion;
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/node/Node.java:702]: This is a nit but I prefer making values final so it is easier to understand the assignment:

```suggestion
        final CompatibleVersion compatibleVersion;
        if (restCompatibilityPlugins.size() > 1) {
            throw new IllegalStateException("Only one RestCompatibilityPlugin is allowed");
        } else if (restCompatibilityPlugins.size() == 1) {
            compatibleVersion = restCompatibilityPlugins.get(0)::getCompatibleVersion;
        } else {
            compatibleVersion = CompatibleVersion.CURRENT_VERSION;
        }
        return compatibleVersion;
```

File: server/src/main/java/org/elasticsearch/plugins/RestCompatibilityPlugin.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * under the License.
+ */
+
+package org.elasticsearch.plugins;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.Nullable;
+
+public interface RestCompatibilityPlugin {
+    Version getCompatibleVersion(@Nullable String acceptHeader, @Nullable String contentTypeHeader, Boolean hasContent);
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/plugins/RestCompatibilityPlugin.java:26]: Please add javadocs

File: server/src/main/java/org/elasticsearch/rest/CompatibleVersion.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * under the License.
+ */
+
+package org.elasticsearch.rest;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.Nullable;
+
+@FunctionalInterface
+public interface CompatibleVersion {
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/rest/CompatibleVersion.java:26]: Please add javadocs here and for the method as well

File: server/src/main/java/org/elasticsearch/rest/CompatibleVersion.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    Version get(@Nullable String acceptHeader, @Nullable String contentTypeHeader, Boolean hasContent);
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/rest/CompatibleVersion.java:27]: can this just be a `boolean`:

```suggestion
    Version get(@Nullable String acceptHeader, @Nullable String contentTypeHeader, boolean hasContent);
```

File: server/src/main/java/org/elasticsearch/rest/RestController.java [hunk 5]
```diff
@@ -295,6 +301,10 @@ private void tryAllHandlers(final RestRequest request, final RestChannel channel
         final String rawPath = request.rawPath();
         final String uri = request.uri();
         final RestRequest.Method requestMethod;
+        //TODO: USAGE_1 now that we have a version we can implement a REST handler that accepts path, method AND version
+//        Version version = request.getCompatibleVersion();
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/rest/RestController.java:305]:
```suggestion
```

File: server/src/main/java/org/elasticsearch/rest/RestHandler.java [hunk 6]
```diff
@@ -89,6 +90,16 @@ default boolean allowsUnsafeBuffers() {
         return Collections.emptyList();
     }
 
+    /**
+     * Returns a version a handler is compatible with.
+     * This version is then used to math a handler with a request that specified a version.
+     * If no version is specified, handler is assumed to be compatible with <code>Version.CURRENT</code>
+     * @return a version
+     */
+    default Version compatibleWithVersion(){
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/rest/RestHandler.java:99]:
```suggestion
    default Version compatibleWithVersion() {
```

File: server/src/main/java/org/elasticsearch/rest/RestRequest.java [hunk 7]
```diff
@@ -173,6 +177,10 @@ public static RestRequest requestWithoutParameters(NamedXContentRegistry xConten
             requestIdGenerator.incrementAndGet());
     }
 
+//    public Version getCompatibleVersion() {
+//        return compatibleVersion.get(header("Accept"), header("Content-Type"), hasContent());
+//    }
+
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/rest/RestRequest.java:183]:
```suggestion
```

File: server/src/test/java/org/elasticsearch/VersionTests.java [hunk 8]
```diff
@@ -369,4 +369,11 @@ public void testUnreleasedVersion() {
         VersionTests.assertUnknownVersion(VERSION_5_1_0_UNRELEASED);
     }
 
+    public void testPreviousVersion(){
+        Version current = Version.CURRENT;
+        assertThat(current.previousMajor(), equalTo(Version.fromString(Version.CURRENT.major-1+".0.0")));
```
Comment (jaymode) [inline server/src/test/java/org/elasticsearch/VersionTests.java:374]:
```suggestion
        assertThat(current.previousMajor(), equalTo(Version.fromString(Version.CURRENT.major - 1 + ".0.0")));
```

File: x-pack/plugin/rest-compatibility/src/main/java/org/elasticsearch/compat/CompatibleApiException.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.compat;
+
+public class CompatibleApiException extends RuntimeException {
```
Comment (jaymode) [inline x-pack/plugin/rest-compatibility/src/main/java/org/elasticsearch/compat/CompatibleApiException.java:9]: Do we need a new exception? Maybe we can just use `ElasticsearchStatusException`?

**Review by jaymode — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/rest/CompatibleVersion.java (hunk 3 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/rest/CompatibleVersion.java:26]: will do

**Comment by pgomulka** [comment]
I will merge that PR into #60516 as most of the discussion was there so far

File: x-pack/plugin/rest-compatibility/src/main/java/org/elasticsearch/compat/CompatibleApiException.java (hunk 9 above)
Comment (pgomulka) [reply x-pack/plugin/rest-compatibility/src/main/java/org/elasticsearch/compat/CompatibleApiException.java:9]: we probably don't need it. will replace with `ElasticsearchStatusException`

**Comment by pgomulka** [comment]
merged into https://github.com/elastic/elasticsearch/pull/60516

**Pull request closed without merging**